File: intellibase-server/src/main/java/com/intellibase/server/domain/dto/ChunkStrategy.java
package com.intellibase.server.domain.dto;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.io.Serializable;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonDeserialize(using = ChunkStrategyDeserializer.class)
public class ChunkStrategy implements Serializable {

    private Integer version;

    private String type;

    private Integer size;

    private Integer overlap;

    private Integer minSize;

    private Boolean normalizeWhitespace;

    /** 是否启用父子分块：子块用于检索，父块上下文用于生成。 */
    private Boolean parentChildEnabled;

    /** 父块目标字符数，建议约 1024~2048 token 对应的字符窗口。 */
    private Integer parentSize;

    /** 子块目标字符数，建议约 256~512 token 对应的字符窗口。 */
    private Integer childSize;

    /** 子块重叠字符数。 */
    private Integer childOverlap;
}

File: intellibase-web/src/utils/chunkStrategy.ts
import type { ChunkStrategy } from '../types';

export const DEFAULT_CHUNK_STRATEGY: ChunkStrategy = {
  version: 2,
  type: 'STRUCTURE_AWARE',
  size: 800,
  overlap: 120,
  minSize: 200,
  normalizeWhitespace: true,
  parentChildEnabled: false,
  parentSize: 1800,
  childSize: 420,
  childOverlap: 80,
};

export const normalizeChunkStrategy = (strategy?: Partial<ChunkStrategy> | null): ChunkStrategy => {
  const size = Math.max(20, Number(strategy?.size ?? DEFAULT_CHUNK_STRATEGY.size));
  const overlap = Math.min(Math.max(0, Number(strategy?.overlap ?? DEFAULT_CHUNK_STRATEGY.overlap)), size - 1);
  const minSize = Math.min(Math.max(1, Number(strategy?.minSize ?? DEFAULT_CHUNK_STRATEGY.minSize)), size);
  const parentSize = Math.max(size, Number(strategy?.parentSize ?? DEFAULT_CHUNK_STRATEGY.parentSize));
  const childSize = Math.max(20, Number(strategy?.childSize ?? DEFAULT_CHUNK_STRATEGY.childSize));
  const childOverlap = Math.min(
    Math.max(0, Number(strategy?.childOverlap ?? DEFAULT_CHUNK_STRATEGY.childOverlap)),
    childSize - 1,
  );

  return {
    version: Number(strategy?.version ?? DEFAULT_CHUNK_STRATEGY.version),
    type: strategy?.type || DEFAULT_CHUNK_STRATEGY.type,
    size,
    overlap,
    minSize,
    normalizeWhitespace: strategy?.normalizeWhitespace ?? DEFAULT_CHUNK_STRATEGY.normalizeWhitespace,
    parentChildEnabled: strategy?.parentChildEnabled ?? DEFAULT_CHUNK_STRATEGY.parentChildEnabled,
    parentSize,
    childSize,
    childOverlap,
  };
};

export const getChunkStrategySummary = (strategy?: Partial<ChunkStrategy> | null) => {
  const normalized = normalizeChunkStrategy(strategy);
  const parentChild = normalized.parentChildEnabled
    ? ` · 父子 ${normalized.parentSize}→${normalized.childSize}/${normalized.childOverlap}`
    : '';
  return `${normalized.type} · ${normalized.size}/${normalized.overlap}${parentChild} · min ${normalized.minSize} · ${
    normalized.normalizeWhitespace ? '规范化空白' : '保留原始空白'
  }`;
};
